File: 958-check-completeness-of-a-binary-tree/958-check-completeness-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        return channaSir(root);
        // return solve(root);
    }
private:
    bool solve(TreeNode* root) {
        queue<TreeNode*> queue;
        queue.push(root);
        
        while(!queue.empty()) {
            int levelSize = queue.size();
            
            while(levelSize--) {
                auto currNode = queue.front();
                queue.pop();
                if(currNode == nullptr){
                    if(!queue.empty() && queue.front() != nullptr){
                        return false;
                    }
                }
                else{
                    queue.push(currNode -> left);
                    queue.push(currNode -> right);
                }
            }
        }
        return true;
    }
    
    bool channaSir(TreeNode* root) {
        queue<TreeNode*> queue;
        queue.push(root);
        
        bool flagNonLeaf = true;
        while(!queue.empty()) {
            auto currNode = queue.front();
            queue.pop();
            
            if(flagNonLeaf == false && (currNode -> left || currNode -> right)) 
                return false;
            if(!currNode -> left && currNode -> right)
                return false;
            
            if((!currNode -> left && !currNode -> right) ||
              (currNode -> left && !currNode -> right))
                flagNonLeaf = false;
            
            if(currNode -> left) queue.push(currNode -> left);
            if(currNode -> right) queue.push(currNode -> right);
        }
        return true;
    }
};

File: 958-check-completeness-of-a-binary-tree/958-check-completeness-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCompleteTree(self, root: Optional[TreeNode]) -> bool:
        queue = deque()
        queue.append(root)
        
        while queue:
            levelSize = len(queue)
            
            for _ in range(levelSize):
                currNode = queue.popleft()
                
                if currNode == None:
                    if len(queue) != 0 and queue[0] != None:
                        return False
                else:
                    queue.append(currNode.left)
                    queue.append(currNode.right)
        return True